File: Phase-2/src/generators/mysql_generator.py
"""MySQL DDL Generator"""
from typing import List
from .base import SQLGenerator
from .type_mapper import map_to_mysql_type
from ..models.metadata import Table, Column

class MySQLGenerator(SQLGenerator):
    """MySQL-specific SQL DDL generator"""
    
    def map_type(self, generic_type: str) -> str:
        """Map generic type to MySQL type"""
        return map_to_mysql_type(generic_type)
    
    def _generate_create_table(self, table: Table) -> str:
        """Generate MySQL CREATE TABLE statement"""
        lines = [f"CREATE TABLE {self._sanitize_name(table.name)} ("]
        
        # Column definitions
        col_defs = []
        for col in table.columns:
            col_def = self._generate_column_definition(col)
            col_defs.append(f"    {col_def}")
        
        # Primary key constraint
        if table.primary_keys:
            pk_cols = [self._sanitize_name(pk.name) for pk in table.primary_keys]
            pk_constraint = f"    PRIMARY KEY ({', '.join(pk_cols)})"
            col_defs.append(pk_constraint)
        
        lines.append(",\n".join(col_defs))
        lines.append(") ENGINE=InnoDB DEFAULT CHARSET=utf8mb4;")
        
        return "\n".join(lines)
    
    def _generate_column_definition(self, column: Column) -> str:
        """Generate column definition"""
        parts = [
            self._sanitize_name(column.name),
            self.map_type(column.data_type)
        ]
        
        # AUTO_INCREMENT for PKs
        if column.is_primary_key and column.data_type.upper() in ['INTEGER', 'INT', 'BIGINT']:
            parts.append("AUTO_INCREMENT")
        
        # NOT NULL
        if not column.is_nullable:
            parts.append("NOT NULL")
        
        # UNIQUE
        if column.is_unique:
            parts.append("UNIQUE")
        
        # DEFAULT
        if column.default_value:
            parts.append(f"DEFAULT {column.default_value}")
        
        return " ".join(parts)
    
    def _generate_foreign_keys(self, table: Table) -> List[str]:
        """Generate ALTER TABLE for foreign keys"""
        statements = []
        
        for idx, fk_col in enumerate(table.foreign_keys, 1):
            if not fk_col.references_table or not fk_col.references_column:
                continue
            
            fk_name = f"fk_{self._sanitize_name(table.name)}_{idx}"
            stmt = (
                f"ALTER TABLE {self._sanitize_name(table.name)} "
                f"ADD CONSTRAINT {fk_name} "
                f"FOREIGN KEY ({self._sanitize_name(fk_col.name)}) "
                f"REFERENCES {self._sanitize_name(fk_col.references_table)}"
                f"({self._sanitize_name(fk_col.references_column)}) "
                f"ON DELETE CASCADE ON UPDATE CASCADE;"
            )
            statements.append(stmt)
        
        return statements
    
    def _generate_indexes(self, table: Table) -> List[str]:
        """Generate CREATE INDEX statements"""
        statements = []
        
        for fk_col in table.foreign_keys:
            idx_name = f"idx_{self._sanitize_name(table.name)}_{self._sanitize_name(fk_col.name)}"
            stmt = f"CREATE INDEX {idx_name} ON {self._sanitize_name(table.name)}({self._sanitize_name(fk_col.name)});"
            statements.append(stmt)
        
        return statements

File: Phase-2/src/generators/MySQLGenerator.js
import { BaseGenerator } from './BaseGenerator.js';
import { mapToMySQLType } from './typeMapper.js';
import path from 'path';
import { writeFile } from '../utils/fileUtils.js';
import logger from '../utils/logger.js';


export class MySQLGenerator extends BaseGenerator {
    //it will now inherit the base-generator..!
    generateCreateTable(table) {
        const lines = [`CREATE TABLE ${this.sanitizeName(table.name)} (`];
        const colDefs = [];


        //column definitions..!
        /**
         * we are taking each column from the table and try to push it to the columnDefinition..!
         */
        for (const col of table.columns) {
            colDefs.push(`${this.generateColumnDefinition(col)}`);
        }

        /**
         * go through each pk in the table object find the primarykeys..!
         * and each pk sanitize it..!
         */
        //primary key constraints...!
        if (table.primaryKeys.length > 0) {
            //if we have the pk make sure it should be santitize..!
            const pkCols = table.primaryKeys.map(pk => this.sanitizeName(pk.name)).join(', ');
            colDefs.push(`    PRIMARY KEY (${pkCols})`);
        }
        lines.push(colDefs.join(',\n'));
        lines.push(') ENGINE=InnoDB DEFAULT CHARSET=utf8mb4;');

        return lines.join('\n');

    }

    generateColumnDefinition(column) {
        const parts = [
            this.sanitizeName(column.name),
            //check if the column dataType is perfect or not..!
            mapToMySQLType(column.dataType),//to get which dataType it is
        ];

        /// AUTO_INCREMENT for PKs
        if (column.isPrimaryKey && ['INTEGER', 'INT', 'BIGINT'].includes(column.dataType.toUpperCase())) {
            parts.push('AUTO_INCREMENT');
            //SO THE auto-increment means that the column will be auto-incremented by 1 each time a new row is inserted.
        }
        //not-null..!
        if (!column.isNullable) {
            parts.push("NOT_NULL");
        }
        //UNIQUE..!
        if (column.isUnique) {
            parts.push('UNIQUE');

        }
        //DEFAULT..!
        if (column.defaultValue) {
            parts.push(`DEFAULT ${column.defaultValue}`);
        }
        return parts.join('');

    }
    generateForeignKeys(table) {
        const statements = [];
        let fkIndex = 1;
        for (const fkCol of table.foreignKeys) {
            //means there is no fk in the reference tabel and in referenceColumn..!
            if (!fkCol.referencesTable || !fkCol.referencesColumn) {
                continue;
            }
            //if there.!
            const fkName = `fk_${this.sanitizeName(table.name)}_${fkIndex}`;
            const stmt = `ALTER TABLE ${this.sanitizeName(table.name)} ` +
                `ADD CONSTRAINT ${fkName} ` +
                `FOREIGN KEY (${this.sanitizeName(fkCol.name)}) ` +
                `REFERENCES ${this.sanitizeName(fkCol.referencesTable)}(${this.sanitizeName(fkCol.referencesColumn)}) ` +
                `ON DELETE CASCADE ON UPDATE CASCADE;`;

            statements.push(stmt);
            fkIndex++;
        }
        return statements.join('\n');
    }
    generateIndexes(table) {
        const statements = [];

        //Its iterating over all the columns in that table(fk-columns..!)
        //Build the index-Name for it..!
        for (const fkCol of table.foreignKeys) {
            const idxName = `idx_${this.sanitizeName(table.name)}_${this.sanitizeName(fkCol.name)}`;
            statements.push(`CREATE INDEX ${idxName} ON ${this.sanitizeName(table.name)}(${this.sanitizeName(fkCol.name)});`);
        }

        return statements;
    }

    async save(fileName) {
        const ddl = this.generateDDL();//call the template..!
        const filePath = path.join(this.outputDir, fileName);//get the file-Path..!
        await writeFile(filePath, ddl);//write the ddl to the file..!
        logger.info({ filePath }, 'MySQL DDL saved successfully');//log the file-Path..!
        return filePath;
    }
}